libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,1107 @@
+
+          const packageDepsFilename: string = `package-deps_${phase.logFilenameIdentifier}.json`;
+          const currentDepsPath: string = path.join(project.projectRushTempFolder, packageDepsFilename);
+          buildCacheContext.currentDepsPath = currentDepsPath;
+
+          let projectDeps: IProjectDeps | undefined;
+          let trackedProjectFiles: string[] | undefined;
+          try {
+            const fileHashes: Map<string, string> | undefined =
+              await createContext.projectChangeAnalyzer._tryGetProjectDependenciesAsync(
+                project,
+                buildCacheTerminal
+              );
+
+            if (fileHashes) {
+              const files: { [filePath: string]: string } = {};
+              trackedProjectFiles = [];
+              for (const [filePath, fileHash] of fileHashes) {
+                files[filePath] = fileHash;
+                trackedProjectFiles.push(filePath);
+              }
+
+              projectDeps = {
+                files,
+                arguments: commandToRun
+              };
+              buildCacheContext.projectDeps = projectDeps;
+            }
+          } catch (error) {
+            // To test this code path:
+            // Delete a project's ".rush/temp/shrinkwrap-deps.json" then run "rush build --verbose"
+            buildCacheTerminal.writeLine(
+              'Unable to calculate incremental state: ' + (error as Error).toString()
+            );
+            buildCacheTerminal.writeLine({
+              text: 'Rush will proceed without incremental execution, caching, and change detection.',
+              foregroundColor: ColorValue.Cyan
+            });
+          }
+
+          if (!projectDeps && buildCacheContext.isSkipAllowed) {
+            // To test this code path:
+            // Remove the `.git` folder then run "rush build --verbose"
+            buildCacheTerminal.writeLine({
+              text: PrintUtilities.wrapWords(
+                'This workspace does not appear to be tracked by Git. ' +
+                  'Rush will proceed without incremental execution, caching, and change detection.'
+              ),
+              foregroundColor: ColorValue.Cyan
+            });
+          }
+
+          // If the deps file exists, remove it before starting execution.
+          FileSystem.deleteFile(currentDepsPath);
+
+          // TODO: Remove legacyDepsPath with the next major release of Rush
+          const legacyDepsPath: string = path.join(project.projectFolder, 'package-deps.json');
+          // Delete the legacy package-deps.json
+          FileSystem.deleteFile(legacyDepsPath);
+
+          // No-op command
+          if (!commandToRun) {
+            // Write deps on success.
+            if (projectDeps) {
+              JsonFile.save(projectDeps, currentDepsPath, {
+                ensureFolderExists: true
+              });
+            }
+            return OperationStatus.NoOp;
+          }
```

Comment:
This isn't necessarily true. Only some Operations have the `commandToRun` field non-empty, but they may still have work that needs to be done.